### PR TITLE
Make it easier to build against custom osrm-backend

### PR DESCRIPTION
**Pull request by springmeyer**
### What Changed

 - centralizes osrm-backend version variable in package.json
 - documents how to build against:
   - a custom osrm-backend version via mason (for portable binaries)
   - or an existing local osrm-backend clone that has already been installed (to iterate fast)
 
**So, no longer do you change the `OSRM_RELEASE` variable in the .travis.yml. Now, instead you change the `osrm_release` variable inside of the `package.json`.**

The idea here is to:

   - Have the `osrm_release` variable just in one place (./bootstrap.sh now pulls from package.json)
   - Avoid using ENV variables and bash `source` which is easy to make mistakes with
   - Make it easier in branch diffs to see the relationship of `osrm-backend` version and `node-osrm` by storing their versions in the same spot.

### GOCHAS

 - If you have the osrm-backend tools on your PATH, they will be used instead of the ones built locally. So, if you are just doing node-osrm development make sure to clean up any stale osrm-backend installs you might have.

### Background

Currently to install node-osrm from binaries you run: `npm install`.

To install from source you type `make`. And to run the tests after a source build: `make test`.

These things will stay the same and they default to building against the osrm-backend `develop` branch.

However when doing development against custom branches of OSRM_BACKEND there were a variety of easy pitfalls.

This branch makes it easier, and documented, how you should:

 - Build against a custom OSRM branch, tag, or gitsha using mason + an automated source build of osrm-backend (done via the ./bootstrap.sh script)
 - Build against a local osrm-backend clone that you've been doing development in

Both these methods are documented in the readme in this pull.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -40,9 +46,12 @@ clean:
...
 	valgrind --leak-check=full node node_modules/.bin/_mocha
 
+# Note: this PATH setting is used to allow the localized tools to be used
+# but your locally installed osrm-backend tool, if on PATH should override
 shm: ./test/data/Makefile
-	$(MAKE) -C ./test/data
-	$(OSRM_DATASTORE) ./test/data/berlin-latest.osrm
+	@PATH="${PATH}:./lib/binding" && echo "*** Using osrm-datastore from `which osrm-datastore` ***"
+	@PATH="${PATH}:./lib/binding" && $(MAKE) -C ./test/data
+	@PATH="${PATH}:./lib/binding" && osrm-datastore ./test/data/berlin-latest.osrm
```
Comment (daniel-j-h) [inline Makefile]: If we add the local binaries to the front of `PATH` does bash prefer the local ones? Would this solve the issue?

File: README.md [hunk 2]
```diff
@@ -159,6 +172,29 @@ Now you can build `node-osrm`:
...
+set the `OSRM_DATA_PATH` variable:
+
+    export OSRM_DATA_PATH=/path/to/osrm-backend/test/data
+
+Then you can run `npm test`.
+
+To recap, here is a full example of building against an osrm-backend that is cloned beside node-osrm but installed into a custom location:
+
+```
+export PATH=/opt/osrm/bin:${PATH}
```
Comment (daniel-j-h) [inline README.md]: Ah here we're doing it! :+1:

File: Makefile [hunk 3]
```diff
@@ -1,32 +1,38 @@
...
+all: build/Release/osrm.node
 
-pkgconfig:
-	@if [[ `which pkg-config` ]]; then echo "Success: Found pkg-config"; else "echo you need pkg-config installed" && exit 1; fi;
+ifeq (, $(shell which pkg-config))
+ $(error "No pkg-config found: please ensure you have pkg-config installed")
+endif
 
-./node_modules/node-pre-gyp:
+./node_modules/nan:
```
Comment (TheMarex) [inline Makefile]: Why is this necessary? I was under the impression that `npm install` would get you this.

Comment (springmeyer) [inline Makefile]: @TheMarex This is not necessary, but an optimization. The goal of the `Makefile` wrapper is to speed up development when only one thing is expected to change: `src/node-osrm.cpp`. Running `npm install` and `make` should have the same result but the former incurs several seconds of wasted time while npm checks the dependency tree and does other fancy stuff and prints thingsto the terminal that are unrelated to C++ development on `src/node-osrm.cpp`. Therefore the Makefile has these shortcuts to quickly avoid re-running `npm install` when it looks like it is unneeded. This is not failsafe: if the first npm install works to install nan but fails to grab another dependency due to some network blip then you might need to do `make clean && make` to get the build working.

**Comment by springmeyer** [comment]
All green now https://travis-ci.org/Project-OSRM/node-osrm/builds/112922109. After merging I will rebase this into the live branches that are pointed at osrm-backend branches which are based on osrm-backend `develop`.
